sqld/src/replication/frame_stream.rs

```diff
@@ -0,0 +1,112 @@
+use std::sync::Arc;
+use std::task::{ready, Poll};
+use std::{pin::Pin, task::Context};
+
+use bytes::Bytes;
+use futures::{Future, Stream};
+
+use super::{FrameNo, LogReadError, ReplicationLogger};
+
+/// Streams pages starting from an offset and only stop if the request frame is older than the
+/// first frame in the current log file.
+pub struct FrameStream {
+    current_frameno: FrameNo,
+    max_available_frame_no: FrameNo,
+    logger: Arc<ReplicationLogger>,
+    state: FrameStreamState,
+}
+
+impl FrameStream {
+    pub fn new(logger: Arc<ReplicationLogger>, current_frameno: FrameNo) -> Self {
+        let max_available_frame_no = *logger.new_frame_notifier.subscribe().borrow();
+        Self {
+            current_frameno,
+            max_available_frame_no,
+            logger,
+            state: FrameStreamState::Init,
+        }
+    }
+
+    fn transition_state_next_frame(&mut self) {
+        if matches!(self.state, FrameStreamState::Closed) {
+            return;
+        }
+
+        let next_frameno = self.current_frameno + 1;
+        let logger = self.logger.clone();
+        let fut = async move {
+            let res = tokio::task::spawn_blocking(move || logger.get_frame(next_frameno)).await;
+            match res {
+                Ok(Ok(frame)) => Ok(frame),
+                Ok(Err(e)) => Err(e),
+                Err(e) => Err(LogReadError::Error(e.into())),
+            }
+        };
+
+        self.state = FrameStreamState::WaitingFrame(Box::pin(fut));
+    }
+}
+
+enum FrameStreamState {
+    Init,
+    /// waiting for new frames to replicate
+    WaitingFrameNo(Pin<Box<dyn Future<Output = anyhow::Result<FrameNo>> + Send + Sync + 'static>>),
+    WaitingFrame(
+        Pin<Box<dyn Future<Output = Result<Bytes, LogReadError>> + Send + Sync + 'static>>,
+    ),
```

Comment:
Nit: why not use `futures::BoxFuture` here? Futures should almost never need to be `Sync`.